### PR TITLE
Implementing bounded message queue #452

### PR DESCRIPTION
**Pull request by avoxm**
Updating BoundedMessageQueue class to act as blocking and bounded
thread-safe collection.

### CODE REVIEW

File: src/core/Akka/Dispatch/MessageQueues/BoundedMessageQueue.cs [hunk 1]
```diff
@@ -7,12 +7,21 @@ namespace Akka.Dispatch.MessageQueues
...
+
+        public BoundedMessageQueue(int boundedCapacity)
+        {
+            _queue = new BlockingCollection<Envelope>(boundedCapacity);
+        }
 
         public void Enqueue(Envelope envelope)
         {
-            _queue.Enqueue(envelope);
+            _queue.Add(envelope);
```
Comment (rogeralsing) [inline src/core/Akka/Dispatch/MessageQueues/BoundedMessageQueue.cs]: Use this overload instead:
https://msdn.microsoft.com/en-us/library/dd287149(v=vs.110).aspx

`TryAdd(Envelope,Timespan)`

And use the `PushTimeout` property.
If I recall correctly, if the timeout occurs, the message should be sent to deadletters, not sure, check Scala impl/docs.

**Comment by rogeralsing** [comment]
This needs to be configurable using the Akka config.

```
    bounded-mailbox {
      mailbox-capacity = 1000
      mailbox-push-timeout-time = 10s
      mailbox-type = "....
```

**Comment by rogeralsing** [comment]
Looks good, just a few minor details that needs fixing

File: src/core/Akka/Dispatch/MessageQueues/BoundedMessageQueue.cs (hunk 1 above)
Comment (avoxm) [inline src/core/Akka/Dispatch/MessageQueues/BoundedMessageQueue.cs]: What is the equivalent class in scala version ?

**Comment by rogeralsing** [comment]
config: https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/dispatch/Mailbox.scala#L642

**Comment by rogeralsing** [comment]
enqueue (post) https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/dispatch/Mailbox.scala#L512

**Comment by rogeralsing** [comment]
@avoxm are you still working on this? otherwise I can fix the last bits myself

**Comment by avoxm** [comment]
@rogeralsing I'll check the scala version and finalize this shortly. Thanks

**Comment by Aaronontheweb** [comment]
@avoxm any reason why this is closed? think all we need is to implement @rogeralsing's feedback

**Comment by avoxm** [comment]
@Aaronontheweb I'll be opening a new pull request to address the notes from @rogeralsing

**Comment by avoxm** [comment]
I guess it would be easier for you to merge this way :)

**Comment by Aaronontheweb** [comment]
Ah, got it :)

**Pull request closed without merging**